### ISSUE TITLE
`ua detach` on Azure 18.04 Pro + FIPS renders VM inaccessible

### ISSUE THREAD

**Issue by cmars** [canonical/ubuntu-advantage-client#1503]
I spun up an Azure Pro + FIPS 18.04 instance and immediately `ua detach`ed (to test a staging contract).

The VM became inaccessible and even restarting it from the console I am unable to SSH back in.

**Comment by blackboxsw** [comment]
Thanks @cmars for the bug. We should add specific logic to better handle disabling fips on a FIPS only image. 

This is a somewhat[ documented issue for manually disabling FIPS in discourse](https://discourse.ubuntu.com/t/ubuntu-advantage-disabling-fips-manually/20738) because Ubuntu PRO FIPS contains only one kernel in the image. So, detaching/disablig fips on this system will result in very bad things. That said, UA client should be smarter here and not allow you to disable if it sees no other bootable kernel available as fallback.

**Comment by blackboxsw** [comment]
Bringing this into our current milestone

**Comment by blackboxsw** [comment]
@cmars thanks for this bug. More details on this specific case:

This is expected on PRO FIPS contracts currently because the contract specifies individual deb packages (instead of ubuntu-fips or ubuntu-aws-fips metapkg) in the `additionalPackages` directives. New UA client from daily PPA attempts to remove any packages listed `additionalPackages` directives because newer contracts will only report the metapkg which *should* be removed as the only way to disable the FIPS services/libraries.    This issue will go away once we can update FIPS PRO contracts when CPC rolls the newest Ubuntu PRO images containing ubuntu-advantage-tools 27.0, when those images are released, we expect to submit a Contract Service API change request to shift PRO image contracts to "ubuntu-aws-fips", "ubuntu-azure-fips" or  "ubuntu-fips" depending on the cloud and Ubuntu release.



For testing and development in the meantime (and switching tokens to staging). You might need to `rm -rf /var/lib/ubuntu-advantage/` so the system doesn't think it is attached and `sudo ua attach <other_token>`

**Comment by davecore82** [comment]
We currently recommend our Ubuntu Pro users to run `sudo ua detach --assume-yes` at the end of the process when they create golden images based on Ubuntu Pro. 

Is there a known issue to not detaching UA during a golden image process?

**Comment by davecore82** [comment]
Actually, users don't even need to run `ua detach` to trigger this issue.  Simply cloning an Ubuntu Pro FIPS 18.04 instance (or creating a Packer image, etc.) in AWS will result in the newly created instances from that AMI to uninstall those FIPS packages upon first boot, leaving the system unbootable as @blackboxsw explained:

```
ubuntu@ip-172-31-62-232:~$ sudo ua status --wait
Removing stale lock file previously held by 1221:ua auto-attach
........................
SERVICE       ENTITLED  STATUS    DESCRIPTION
esm-infra     yes       disabled  UA Infra: Extended Security Maintenance (ESM)
fips          yes       disabled  NIST-certified FIPS modules
fips-updates  no        —         Uncertified security updates to FIPS modules
livepatch     yes       n/a       Canonical Livepatch service

NOTICES
Operation in progress: ua auto-attach

Enable services with: ua enable <service>

                Account: <redacted>
           Subscription: <redacted>
            Valid until: 9999-12-31 00:00:00
Technical support level: essential

ubuntu@ip-172-31-62-232:~$ dpkg -l | grep -i fips
rc  fips-initramfs                           0.0.10                                          amd64        FIPS 140-2 kernel tests
ii  kcapi-tools                              1.0.3-2fips3                                    amd64        Command-line tools for Linux Kernel Crypto API
ii  libcharon-standard-plugins               5.6.2-1ubuntu2.fips.2.5.1                       amd64        strongSwan charon library (standard plugins)
ii  libkcapi1:amd64                          1.0.3-2fips3                                    amd64        Linux Kernel Crypto API User Space Interface Library
ri  libssl1.1:amd64                          1.1.1-1ubuntu2.fips.2.1~18.04.3.1               amd64        Secure Sockets Layer toolkit - shared libraries
ii  libstrongswan                            5.6.2-1ubuntu2.fips.2.5.1                       amd64        strongSwan utility and crypto library
rc  libstrongswan-standard-plugins           5.6.2-1ubuntu2.fips.2.5.1                       amd64        strongSwan utility and crypto library (standard plugins)
ii  linux-aws-fips-headers-4.15.0-2000       4.15.0-2000.4                                   all          Header files related to Linux kernel version 4.15.0
rH  linux-headers-4.15.0-2000-aws-fips       4.15.0-2000.4                                   amd64        Linux kernel headers for version 4.15.0 on 64 bit x86 SMP
ii  linux-image-4.15.0-2000-aws-fips         4.15.0-2000.4                                   amd64        Signed FIPS 140-2 kernel image for AWS
ii  linux-image-hmac-4.15.0-2000-aws-fips    4.15.0-2000.4                                   amd64        HMAC file for linux kernel image for version 4.15.0-2000
ii  linux-modules-4.15.0-2000-aws-fips       4.15.0-2000.4                                   amd64        Linux kernel extra modules for version 4.15.0 on 64 bit x86 SMP
rc  linux-modules-extra-4.15.0-2000-aws-fips 4.15.0-2000.4                                   amd64        Linux kernel extra modules for version 4.15.0 on 64 bit x86 SMP
ri  openssh-client                           1:7.9p1-10~ubuntu18.04.fips.0.2                 amd64        secure shell (SSH) client, for secure access to remote machines
ri  openssh-client-hmac:amd64                1:7.9p1-10~ubuntu18.04.fips.0.2                 amd64        FIPS HMAC integrity check files for secure shell (SSH) client.
rc  openssh-server                           1:7.9p1-10~ubuntu18.04.fips.0.2                 amd64        secure shell (SSH) server, for secure access from remote machines
ri  openssh-sftp-server                      1:7.9p1-10~ubuntu18.04.fips.0.2                 amd64        secure shell (SSH) sftp server module, for SFTP access from remote machines
ri  openssl                                  1.1.1-1ubuntu2.fips.2.1~18.04.3.1               amd64        Secure Sockets Layer toolkit - cryptographic utility
ii  strongswan-charon                        5.6.2-1ubuntu2.fips.2.5.1                       amd64        strongSwan Internet Key Exchange daemon
ii  strongswan-libcharon                     5.6.2-1ubuntu2.fips.2.5.1                       amd64        strongSwan charon library
ii  strongswan-starter                       5.6.2-1ubuntu2.fips.2.5.1                       amd64        strongSwan daemon starter and configuration file parser

ubuntu@ip-172-31-62-232:~$ sudo reboot
Connection to 18.207.234.203 closed by remote host.
Connection to 18.207.234.203 closed.

# can't ssh back in instance
```

**Comment by blackboxsw** [comment]
Thanks again @davecore82 cloning an image also exercises `ua detach` logic if the booting instance represents a new instance-id (which happens across an image clone on most clouds).   That said, we have just released version ubuntu-advantage-tools version 27.0  in AWS and Azure Ubuntu PRO and  Ubuntu PRO FIPS images as of 06/11/2021. With this release, contract changes took place which limit `ua detach` to only remove ubuntu-fips, ubuntu-aws-fips or ubuntu-azure-fips metapkg instead of removing openssl/libssl and rendering the system inaccessible.
